compiler/src/iree/compiler/Codegen/Transforms/Patterns.h

```diff
@@ -0,0 +1,38 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_CODEGEN_TRANSFORMS_PATTERNS_H_
+#define IREE_COMPILER_CODEGEN_TRANSFORMS_PATTERNS_H_
+
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+
+namespace mlir::iree_compiler {
+
+/// Fold `tensor.pad(cst, tensor.extract*(linalg.fill(cst)))` into
+/// `linalg.fill(cst, empty)` when the padding constant and the fill constant
+/// are the same.
+/// This seems generally desirable as a folding but may be too intrusive, so we
+/// only apply it selectively for now.
+// TODO: atm hardcoded on linalg.fill but we could take any result of any
+// generic that yields a constant in that result.
+struct FoldFillIntoPad : public OpRewritePattern<tensor::PadOp> {
+  using OpRewritePattern<tensor::PadOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensor::PadOp padOp,
+                                PatternRewriter &rewriter) const;
+};
+
+/// Pattern to rewrite tensor.empty to tensor.alloc.
+struct EmptyTensorLoweringPattern : public OpRewritePattern<tensor::EmptyOp> {
+  using OpRewritePattern<tensor::EmptyOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::EmptyOp op,
+                                PatternRewriter &rewriter) const;
+};
```

Comment:
Is there a reason to add the pattern structs here instead of populate patterns? I just haven't seen this before.

Reply:
It is hard to categorize the populate methods because each method would only populate one pattern. We have such structure in upstream for a long time, so I think it is fine. E.g., https://github.com/llvm/llvm-project/blob/396cdabe47f3596464b289d0937c0066f50a0ac6/mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h#L1247-L1463

Reply:
oh cool, makes sense to me.